<html>
    <head>
        <title>Two Sample Logos - Help</title>
        <link rel="stylesheet" href="tsl.css" type="text/css">
        <link rel="shortcut icon" href="images/favicon.ico">
        <meta http-equiv="Content-Type" content="text/html;charset=UTF-8">
    </head>
    <body>
        <table width="100%" cellpadding="0" cellspacing="6" border="0" align="center">
        <tr>
        <td align="left" valing="bottom"><a href="index.html"><img src="images/tsl.gif" width="206" height="60" border="0" alt="Two Sample Logos" hspace="0" vspace="6"></a></td>
        <td align="right" valign="bottom">
<a href="index.html">Home</a> |
<a href="../cgi-bin/tsl/tsl.cgi">Create TSL</a> | 
<a href="examples.html">Examples</a> |
Help</td>
        </tr>
        <tr>
        <td colspan="2">

<h4>Options</h4>

<a name="sequence"></a><h3>Positive and negative sample</h3>
<p>
Two Sample Logo calculates statistical significance of the relative 
position-specific symbol frequencies between two sets of aligned 
sequences. For example, sequences that are known to share a sequence 
motif may be locally aligned including positions upstream or downstream 
from the motif. All aligned sequences in both samples are required to 
be of the same length, so dash characters ("-") should be used to pad 
the positions in case some sequences are shorter.
</p>

<p>
Sequences that contain a motif and at the same time have a certain 
functional property (say, protein modification sites or transcription 
factor binding regions) constitute a <b>positive sample</b>. Sequences 
that contain the motif and at the same time do not have the functional 
property constitute the <b>negative sample</b>. The distinction between
the samples does not necessarily have to be based on the presence and 
absence of a functional property: as long as there is a clear way of 
interpreting the data, any pair of sets of aligned sequences can be 
used as positive and negative.
</p>

<p>
Sequences can be entered as flat files, or in <a href="http://www.ncbi.nlm.nih.gov/blast/fasta.shtml">FASTA</a> 
or <a href="http://www.ebi.ac.uk/help/ClustalW_frame.html">ClustalW</a> formats.
</p>

<a name="seq_type"></a><h3>Sequence type</h3>
<p>
Either amino acid or nucleotide. If amino acid option is selected, all
symbols other than the standard 20 amino acid single-letter codes will
be replaced with dashes and will not be a part of the statistics. Likewise, 
if nucleotide option is selected, all symbols other than the A, C, G, 
T, and U will be replaced with dashes. 
</p>

<a name="test"></a><h3>Statistical tests</h3>
<p>
Two Sample Logo supports two types of statistical tests: 
<ol>
<li>two sample t-test</li>
<li>binomial test</li>
</ol>
Both tests estimate the p-value of the null hypothesis that both positive
and negative samples were generated by the same distribution, however 
they use different assumptions. 
</p>

<font color="black"><b>Two sample t-test</b></font>

<p>
Frequently used statistical procedure that tests whether two samples
were generated by the same Gaussian distribution. The assumptions of the
t-test are that all observations are independent and that the standard
deviations for both samples are identical, then it checks the equality
of means (Hogg and Craig, 1994).
</p>


<font color="black"><b>Binomial test</b></font>

<p>
Consider two 0-1 samples S<sub>1</sub> and S<sub>2</sub> of sizes n<sub>1</sub> 
and n<sub>2</sub> respectively, in which symbol 1 occurred k<sub>1</sub> times 
in S<sub>1</sub> and k<sub>2</sub> times in S<sub>2</sub>. Let us also assume 
that the test statistic is the absolute difference of symbol’s relative 
frequencies, i.e. θ  = |k<sub>1</sub>/n<sub>1</sub> – k<sub>2</sub>/n<sub>2</sub>|. 
The binomial test calculates the probability that a difference ≥θ  for the two 
samples of sizes n1 and n2 randomly drawn from the underlying null distribution 
could occur by chance alone. Since, according to the null model, both samples 
are independent and identically distributed, an unbiased estimate of the
probability of success p of the underlying binomial distribution is calculated 
as the relative frequency of occurrence of a symbol when S<sub>1</sub> and S<sub>2</sub> 
are concatenated, i.e. p = (k<sub>1</sub> + k<sub>2</sub>)/(n<sub>1</sub> + 
n<sub>2</sub>). The achieved significance level P of the null hypothesis is 
then the probability that the difference ≥θ will be observed between the 
estimated success probabilities in the two samples of sizes n<sub>1</sub> and 
n<sub>2</sub> randomly drawn from the underlying distribution. It is 
calculated as:
</p>

<img src="images/binomial.png" width="402" height="97" border="0">


<a name="p_value"></a><h3>P-value</h3>
<p>
P-value is defined as the lowest significance level at which the null 
hypothesis can be rejected. In the case of two sample logos, null 
hypothesis assumes that each symbol at each position in both samples is 
generated according to the same probability distribution. Based on the 
null hypothesis, p-value is calculated as the probability that the test 
statistic as extreme or more extreme than in the original samples can 
occur by chance alone. Here, the test statistic is the absolute value of 
the difference in relative frequencies between positive and negative 
samples. Since in most cases this probability cannot be calculated 
exactly, p-value is only approximated.
</p>



<a name="conserved"></a><h3>Show conserved residues</h3>
<p>
Because conserved motifs will not be enriched nor depleted in the positive
sample in comparison to the negative sample (the difference of their 
relative frequencies will be zero), by default they will not be displayed 
in the logo. Checking this option forces the software to show conserved
residues.
</p>



<a name="fixed"></a><h3>Fixed height symbols</h3>
<p>
When this option is checked, all enriched and depleted symbols will have the 
same height. When it is not checked, the height of the symbols will be 
proportional to the difference of relative frequencies of corresponding
residues at a given position in the positive and negative sample.
</p>



<a name="bonferroni"></a><h3>Bonferroni correction</h3>
<p>
A correction of the p-value in cases when multiple dependent or
independent hypotheses are tested. See (Weisstein) for <a href="http://mathworld.wolfram.com/BonferroniCorrection.html">
details</a>.
</p>

<h4>Advanced options</h4>

<a name="title"></a><h3>Title</h3>
<p>
Sets up the title for the two sequence logo.
</p>



<a name="range"></a><h3>Logo range</h3>
<p>
Limits the analysis to the specified colums in the samples of aligned
sequences.
</p>



<a name="firstnum"></a><h3>First position index</h3>
<p>
Index assigned to the first symbol in the logo. For example, if the sample
is a 25 residue-long window centered around an active site, first position
symbol should be -12: then the active site will have index 0, and the last
symbol will be indexed as +12. The default value is 1.
</p>



<a name="xaxis"></a><h3>Show X-axis indexes</h3>
<p>
Show residue indexes on the X-axis.
</p>



<a name="yaxis"></a><h3>Show Y-axis labels</h3>
<p>
Shows labels "enriched" and "depleted" next to the Y-axis.
</p>



<h4>Output options</h4>

<a name="format"></a><h3>Output format</h3>
<p>
Two Sample Logo supports Encapsulated PostScript (EPS), Portable Document
Format (PDF), Graphics Interchange Format (GIF) and Portable Network
Graphics (PNG). 
</p>



<a name="size"></a><h3>Output size</h3>
<p>
Height and width of the output image, in pixels, centimeters or inches.
</p>



<a name="resolution"></a><h3>Resolution</h3>
<p>
Sets up the image resolution. Applicable to bitmaps only (GIF and PNG).
</p>



<a name="antialias"></a><h3>Antialiasing</h3>
<p>
Turns antialiasing on or off.
</p>



<a name="box"></a><h3>Boxed image</h3>
<p>
If this option is checked, letters in the output will be inscribed in
bounding boxes. 
</p>



<a name="outline"></a><h3>Outlined symbols</h3>
<p>
If this option is checked, letters in the output will be only outlined 
(and not filled).
</p>



<h4>Color schemes</h4>

<a name="bw"></a><h3>Black and white</h3>

<p>
All symbols are written in black type against a white background.
</p>


<a name="amino_weblogo"></a><h3>WebLogo default colors</h3>

<table width="400" cellpadding=2 cellspacing=0 border=1>
<tr><td width="100">Amino Acid</td>
<td width="100">Color Name</td>
<td width="100">RGB</td>
<td width="100">Hexadecimal</td></tr>

<tr><td width="100"><font color="#00CC00"><b>G, S, T, Y, C</b></td>
<td width="100"><font color="#00CC00"><b>green</b></td>
<td width="100"><TT>[0,204,0]</TT></td>
<td width="100">00CC00</td></tr>

<tr><td width="100"><font color="#CC00CC"><b>N, Q</b></td>
<td width="100"><font color="#CC00CC"><b>purple</b></td>
<td width=100><TT>[204,0,204]</TT></td>
<td width="100">CC00CC</td></tr>

<tr><td width="100"><font color="#0000CC"><b>K, R, H</b></td>
<td width="100"><font color="#0000CC"><b>blue</b></td>
<td width="100"><TT>[0,204,0]</TT></td>
<td width="100">0000CC</td></tr>

<tr><td width="100"><font color="#CC0000"><b>D, E</b></td>
<td width="100"><font color="#CC0000"><b>red</b></td>
<td width="100"><TT>[204,0,0]</TT></td>
<td width="100">CC0000</td></tr>

<tr><td width="100"><font color="#000000"><b>P,A,W,F,L,I,M,V</b></td>
<td width="100"><font color="#000000"><b>black</b></td>
<td width="100"><TT>[0,0,0]</TT></td>
<td width="100">000000</td></tr>
</table><br><br>



<a name="nucleo_weblogo"></a>

<table width="400" cellpadding=2 cellspacing=0 border=1>
<tr><td width="100">Nucleotide</td>
<td width="100">Color Name</td>
<td width="100">RGB</td>
<td width="100">Hexadecimal</td></tr>

<tr><td width="100"><font color="#FFBB00"><b>G</b></td>
<td width="100"><font color="#FFBB00"><b>orange</b></td>
<td width="100"><TT>[255,187,0]</TT></td>
<td width="100">FFBB00</td></tr>

<tr><td width="100"><font color="#CC0000"><b>T, U</b></td>
<td width="100"><font color="#CC0000"><b>red</b></td>
<td width=100><TT>[204,0,0]</TT></td>
<td width="100">CC0000</td></tr>

<tr><td width="100"><font color="#0000CC"><b>C</b></td>
<td width="100"><font color="#0000CC"><b>blue</b></td>
<td width="100"><TT>[0,0,204]</TT></td>
<td width="100">0000CC</td></tr>

<tr><td width="100"><font color="#00CC00"><b>A</b></td>
<td width="100"><font color="#00CC00"><b>green</b></td>
<td width="100"><TT>[0,204,0]</TT></td>
<td width="100">00CC00</td></tr>
</table>



<a name="amino_shapley"></a><h3>Shapley color table for amino acids</h3>

<table width="400" cellpadding=2 cellspacing=0 border=1>
<tr><td width="100">Amino Acids</td>
<td width="100">Color Name</td>
<td width="100">RGB</td>
<td width="100">Hexadecimal</td></tr>

<tr><td width="100"><font color="#A00042"><b>D, T</b></td>
<td width="100"><font color="#A00042"><b>dark red</b></td>
<td width="100"><TT>[160,0,66]</TT></td>
<td width="100">A00042</td></tr>

<tr><td width="100"><font color="#660000"><b>E</b></td>
<td width="100"><font color="#660000"><b>red-brown</b></td>
<td width="100"><TT>[102,0,0]</TT></td>
<td width="100">660000</td></tr>

<tr><td width="100"><font color="#FFFF70"><b>C</b></td>
<td width="100"><font color="#FFFF70"><b>bright yellow</b></td>
<td width="100"><TT>[255,255,112]</TT></td>
<td width="100">FFFF70</td></tr>

<tr><td width="100"><font color="#B8A042"><b>M, Y</b></td>
<td width="100"><font color="#B8A042"><b>dark yellow</b></td>
<td width="100"><TT>[184,160,66]</TT></td>
<td width="100">B8A042</td></tr>

<tr><td width="100"><font color="#4747B8"><b>K</b></td>
<td width="100"><font color="#4747B8"><b>blue</b></td>
<td width="100"><TT>[71,71,184]</TT></td>
<td width="100">4747B8</td></tr>

<tr><td width="100"><font color="#00007C"><b>R</b></td>
<td width="100"><font color="#00007C"><b>dark blue</b></td>
<td width="100"><TT>[0,0,124]</TT></td><td width="100">00007C</td></tr>

<tr><td width="100"><font color="#FF4C4C"><b>S, Q</b></td>
<td width="100"><font color="#FF4C4C"><b>orange</b></td>
<td width="100"><TT>[255,76,76]</TT></td>
<td width="100">FF4C4C</td></tr>

<tr><td width="100"><font color="#534C42"><b>F, P, W</b></td>
<td width="100"><font color="#534C42"><b>dark grey</b></td>
<td width="100"><TT>[83,76,66]</TT></td>
<td width="100">534C42</td></tr>

<tr><td width="100"><font color="#FF7C70"><b>N</b></td>
<td width="100"><font color="#FF7C70"><b>flesh</b></td>
<td width="100"><TT>[255,124,112]</TT></td>
<td width="100">FF7C70</td></tr>

<tr><td width="100"><font color="#EBEBEB"><b>G, V</b></td>
<td width="100"><font color="#EBEBEB"><b>light grey</b></td>
<td width="100"><TT>[235,235,235]</TT></td>
<td width="100">EBEBEB</td></tr>

<tr><td width="100"><font color="#004C00"><b>I</b></td>
<td width="100"><font color="#004C00"><b>dark green</b></td>
<td width="100"><TT>[0,76,0]</TT></td>
<td width="100">004C00</td></tr>

<tr><td width="100"><font color="#455E45"><b>L</b></td>
<td width="100"><font color="#455E45"><b>grey-green</b></td>
<td width="100"><TT>[69,94,69]</TT></td>
<td width="100">455E45</td></tr>

<tr><td width="100"><font color="#8CFF8C"><b>A</b></td>
<td width="100"><font color="#8CFF8C"><b>light green</b></td>
<td width="100"><TT>[140,255,140]</TT></td>
<td width="100">8CFF8C</td></tr>

<tr><td width="100"><font color="#7070FF"><b>H</b></td>
<td width="100"><font color="#7070FF"><b>pale blue</b></td>
<td width="100"><TT>[112,112,255]</TT></td>
<td width="100">7070FF</td></tr>
</table>

<p>
In the original Shapley scheme, G and V were color-coded as white. Since this
would render them invisible against a white background, their color has been
changed to light grey.
</p>

<a name="nucleo_shapley"><h3>Shapley color table for nucleotides</h3>

<table width="400" cellpadding=2 cellspacing=0 border=1>
<tr><td width="100">Nucleotide</td>
<td width="100">Color Name</td>
<td width="100">RGB</td>
<td width="100">Hexadecimal</td></tr>

<tr><td width="100"><font color="#A0A0FF"><b>A</b></td>
<td width="100"><font color="#A0A0FF"><b>light blue</b></td>
<td width="100"><TT>[160,160,255]</TT></td>
<td width="100">A0A0FF</td></tr>

<tr><td width="100"><font color="#FF8C4B"><b>C</b></td>
<td width="100"><font color="#FF8C4B"><b>orange</b></td>
<td width="100"><TT>[255,140,75]</TT></td>
<td width="100">FF8C4B</td></tr>

<tr><td width="100"><font color="#FF7070"><b>G</b></td>
<td width="100"><font color="#FF7070"><b>light red</b></td>
<td width="100"><TT>[255,112,112]</TT></td>
<td width="100">FF7070</td></tr>

<tr><td width="100"><font color="#A0FFA0"><b>T</b></td>
<td width="100"><font color="#A0FFA0"><b>light green</b></td>
<td width="100"><TT>[160,255,160]</TT></td>
<td width="100">A0FFA0</td></tr>

<tr><td width="100"><font color="#B8B8B8"><b>U</b></td>
<td width="100"><font color="#B8B8B8"><b>dark grey</b></td>
<td width="100"><TT>[184,184,184]</TT></td>
<td width="100">B8B8B8</td></tr>
</table>



<a name="amino_colors"><h3>Amino Colors</h3>

<table width="400" cellpadding=2 cellspacing=0 border=1>
<tr><td width="100">Amino Acid</td>
<td width="100">Color Name</td>
<td width="100">RGB</td>
<td width="100">Hexadecimal</td></tr>

<tr><td width="100"><font color="#E60A0A"><b>D, E</b></td>
<td width="100"><font color="#E60A0A"><b>bright red</b></td>
<td width="100"><TT>[230,10,10]</TT></td>
<td width="100">E60A0A</td></tr>

<tr><td width="100"><font color="#E6E600"><b>C, M</b></td>
<td width="100"><font color="#E6E600"><b>yellow</b></td>
<td width=100><TT>[230,230,0]</TT></td>
<td width="100">E6E600</td></tr>

<tr><td width="100"><font color="#145AFF"><b>K, R</b></td>
<td width="100"><font color="#145AFF"><b>blue</b></td>
<td width="100"><TT>[20,90,255]</TT></td>
<td width="100">145AFF</td></tr>

<tr><td width="100"><font color="#FA9600"><b>S, T</b></td>
<td width="100"><font color="#FA9600"><b>orange</b></td>
<td width="100"><TT>[250,150,0]</TT></td>
<td width="100">FA9600</td></tr>

<tr><td width="100"><font color="#3232AA"><b>F, Y</b></td>
<td width="100"><font color="#3232AA"><b>mid blue</b></td>
<td width="100"><TT>[50,50,170]</TT></td>
<td width="100">3232AA</td></tr>

<tr><td width="100"><font color="#00DCDC"><b>N, Q</b></td>
<td width="100"><font color="#00DCDC"><b>cyan</b></td>
<td width="100"><TT>[0,220,220]</TT></td>
<td width="100">00DCDC</td></tr>

<tr><td width="100"><font color="#EBEBEB"><b>G</b></td>
<td width="100"><font color="#EBEBEB"><b>light grey</b></td>
<td width="100"><TT>[235,235,235]</TT></td>
<td width="100">EBEBEB</td></tr>

<tr><td width="100"><font color="#0F820F"><b>L, V, I</b></td>
<td width="100"><font color="#0F820F"><b>green</b></td>
<td width="100"><TT>[15,130,15]</TT></td>
<td width="100">0F820F</td></tr>

<tr><td width="100"><font color="#C8C8C8"><b>A</b></td>
<td width="100"><font color="#C8C8C8"><b>dark grey</b></td>
<td width="100"><TT>[200,200,200]</TT></td>
<td width="100">C8C8C8</td></tr>

<tr><td width="100"><font color="#B45AB4"><b>W</b></td>
<td width="100"><font color="#B45AB4"><b>pink</b></td>
<td width="100"><TT>[180,90,180]</TT></td>
<td width="100">B45AB4</td></tr>

<tr><td width="100"><font color="#8282D2"><b>H</b></td>
<td width="100"><font color="#8282D2"><b>pale blue</b></td>
<td width="100"><TT>[130,130,210]</TT></td>
<td width="100">8282D2</td></tr>

<tr><td width="100"><font color="#DC9682"><b>P</b></td>
<td width="100"><font color="#DC9682"><b>flesh</b></td>
<td width="100"><TT>[220,150,130]</TT></td>
<td width="100">DC9682</td></tr>
</table>


<a name="amino_charge"></a><h3>Charge</h3>
<p>
Positively charged residues (<font color="blue">K, R, H</font>) are colored 
blue, and negatively charged residues (<font color="red">D, E</font>) are 
colored red; all neutral residues are colored black. 
</p>



<a name="amino_hydro"></a><h3>Hydrophobicity</h3>
<p>
Hydrophobic residues (<font color="cyan">A, F, G, I, L, P, V, W, Y</font>) are 
cyan colored, while the remaining hydrophilic residues are colored black. This 
classification was based on (Eisenberg, 1984).
</p>




<a name="amino_surface"></a><h3>Surface exposure</h3>
<p>
Surface exposed residues (<font color="orange">D, E, H, K, N, P, Q, R, S, 
T, Y</font>) are colored orange, and burried residues 
(<font color="black">A, C, F, G, I, L, M, V, W</font>) are colored 
black. This classification was based on (Janin, 1979).
</p>




<a name="amino_flex"></a><h3>Flexibility</h3>
<p>
High flexibility residues (<font color="red">D, E, K, N, P, Q, R, S</font>)
are colored red, whereas low flexibility residues (<font color="green">A, 
C, F, G, H, I, L, M, T, V, W, Y</font>) are colored green. This 
classification was based on (Vihinen et al., 1994).
</p>





<a name="amino_disorder"></a><h3>Disorder</h3>
<p>
Disorder-promoting residues (<font color="red">A, R, S, Q, E, G, K, P</font>)
are colored red, order-promoting residues (<font color="blue">N, C, I, L, F, 
W, Y, V</font>) are colored blue, and disorder-order neutral residues 
(D, H, M, T) are colored black. This classification was based on (Dunker et al., 2001).</p>



<a name="user_colors"><h3>User defined color scheme</h3>

<p>
This option allows you to specify a new color mapping using the set of standard predefined colors listed in the following table:
</p>


<table width="300" cellpadding=2 cellspacing=0 border=1>
<tr><td width="100" height="20">Color</td>
<td width="100" height="20">RGB</td>
<td width="100" height="20">Hexadecimal</td></tr>

<tr><td width="100" height="18"><font color="#0000FF"><TT><b>blue</b></TT></FONT></td>
<td width="100" height="18"><TT>[0,0,255]</TT></td>
<td width="100" height="18">0000FF</td></tr>

<tr><td width="100" height="18"><font color="#000000"><TT><b>black</b></TT></FONT></td>
<td width="100" height="18"><TT>[0,0,0]</TT></td>
<td width="100" height="18">000000</td></tr>

<tr><td width="100" height="18"><font color="#00FFFF"><TT><b>cyan</b></TT></FONT></td>
<td width="100" height="18"><TT>[0,255,255]</TT></td>
<td width="100" height="18">00FFFF</td></tr>

<tr><td width="100" height="18"><font color="#00FF00"><TT><b>green</b></TT></FONT></td>
<td width="100" height="18"><TT>[0,255,0]</TT></td>
<td width="100" height="18">00FF00</td></tr>

<tr><td width="100" height="18"><font color="#2E8B57"><TT><b>greenblue</b></TT></FONT></td>
<td width="100" height="18"><TT>[46,139,87]</TT></td>
<td width="100" height="18">2E8B57</td></tr>

<tr><td width="100" height="18"><font color="#FF00FF"><TT><b>magenta</b></TT></FONT></td>
<td width="100" height="18"><TT>[0,255,0]</TT></td>
<td width="100" height="18">FF00FF</td></tr>

<tr><td width="100" height="18"><font color="#FFA500"><TT><b>orange</b></TT></FONT></td>
<td width="100" height="18"><TT>[255,165,0]</TT></td>
<td width="100" height="18">FFA500</td></tr>

<tr><td width="100" height="18"><font color="#A020F0"><TT><b>purple</b></TT></FONT></td>
<td width="100" height="18"><TT>[160,32,240]</TT></td>
<td width="100" height="18">A020F0</td></tr>

<tr><td width="100" height="18"><font color="#FF0000"><TT><b>red</b></TT></FONT></td>
<td width="100" height="18"><TT>[255,0,0]</TT></td>
<td width="100" height="18">FF0000</td></tr>

<tr><td width="100" height="18"><font color="#FF4500"><TT><b>redorange</b></TT></FONT></td>
<td width="100" height="18"><TT>[255,69,0]</TT></td>
<td width="100" height="18">FF4500</td></tr>

<tr><td width="100" height="18"><font color="#EE82EE"><TT><b>violet</b></TT></FONT></td>
<td width="100" height="18"><TT>[238,130,238]</TT></td>
<td width="100" height="18">EE82EE</td></tr>

<tr><td width="100" height="18"><font color="#FFFFFF"><TT><b>white</b></TT></FONT></td>
<td width="100" height="18"><TT>[255,255,255]</TT></td>
<td width="100" height="18">FFFFFF</td></tr>

<tr><td width="100" height="18"><font color="#FFFF00"><TT><b>yellow</b></TT></FONT></td>
<td width="100" height="18"><TT>[255,255,0]</TT></td>
<td width="100" height="18">FFFF00</td></tr>
</table>

<p>
Any symbol not explicitly assigned to a color will default to black.
</p>


<h4>References</h4>

<ul>

<li>Dunker, A.K., Lawson, J.D., Brown, C.J., Williams, R.M., Romero, P., Oh, J.S., Oldfield, C.J., Campen, A.M., Ratliff, C.M., Hipps, K.W. et al. (2001) "Intrinsically disordered protein." <u>J. Mol. Graph. Model.</u>, 19, 26-59. <a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&db=pubmed&dopt=Abstract&list_uids=11381529&query_hl=5"><img src="images/pubmed.gif" width="47" height="15" border="0"></a></li>

<li>Eisenberg, D. (1984) "Three-dimensional structure of membrane and surface proteins." <u>Annu. Rev. Biochem.</u>, 53, 595-623. <a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&db=pubmed&dopt=Abstract&list_uids=6383201&query_hl=10"><img src="images/pubmed.gif" width="47" height="15" border="0"></a></li>

<li>Hogg, R. V. and Craig A. (1994) <u>Introduction to Mathematical Statistics</u>, 5th edition, Prentice Hall.</li>

<li>Janin, J. (1979) "Surface and inside volumes in globular proteins." <u>Nature</u>, 277, 491-492. <a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&db=pubmed&dopt=Abstract&list_uids=763335&query_hl=12"><img src="images/pubmed.gif" width="47" height="15" border="0"></a></li>

<li>Vihinen, M., Torkkila, E. and Riikonen, P. (1994) "Accuracy of protein flexibility predictions." <u>Proteins</u>, 19, 141-149. <a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&db=pubmed&dopt=Abstract&list_uids=8090708&query_hl=14"><img src="images/pubmed.gif" width="47" height="15" border="0"></a></li>

<li>Weisstein, E. W. "Bonferroni Correction." From <u>MathWorld</u> - A Wolfram Web Resource. <a href="http://mathworld.wolfram.com/BonferroniCorrection.html">http://mathworld.wolfram.com/BonferroniCorrection.html</a></li>

</ul>

        <hr noshade size="3">

        </td>
        </tr>
        </table>
    </body>
</html>
